### PR TITLE
Conversion of `pointer_object_exprt` and `pointer_offset_exprt` for new SMT backend

### PR DESCRIPTION
**Pull request by NlightNFotis**
This PR adds the implementation (along with unit and regression tests)
for the conversion of `pointer_object_exprt`s and `pointer_offset_exprt`s for the new SMT backend.

(Because of some limitations for now in our support for arrays, pointer arithmetic, `pointer_offset_exprt`s are covered by unit tests only) 

- [ ] Each commit message has a non-empty body, explaining why the change was made.
- [ ] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [ ] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [ ] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [ ] My commit message includes data points confirming performance improvements (if claimed).
- [ ] My PR is restricted to a single feature or bugfix.
- [ ] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

### CODE REVIEW

File: unit/solvers/smt2_incremental/convert_expr_to_smt.cpp [hunk 1]
```diff
@@ -1186,3 +1186,103 @@ TEST_CASE(
...
   }
 }
+
+TEST_CASE(
+  "expr to smt conversion for pointer object expression",
+  "[core][smt2_incremental]")
+{
+  // Prelude similar to the one used for address-of expression conversion.
+
+  // The config lines are necessary to ensure that pointer width in configured.
```
Comment (TGWDB) [inline unit/solvers/smt2_incremental/convert_expr_to_smt.cpp:1195]: in -> is

**Review by TGWDB — APPROVED** [review]

File: src/solvers/smt2_incremental/smt_bit_vector_theory.cpp [hunk 2]
```diff
@@ -69,7 +69,6 @@ std::vector<smt_indext> smt_bit_vector_theoryt::extractt::indices() const
 
 void smt_bit_vector_theoryt::extractt::validate(const smt_termt &operand) const
 {
-  PRECONDITION(i >= j);
```
Comment (thomasspriggs) [inline src/solvers/smt2_incremental/smt_bit_vector_theory.cpp]: :no_entry_sign: This precondition should not be removed as far as I can see. Use of `extract` is not being added in this commit and you have made no argument for its removal.

File: src/solvers/smt2_incremental/object_tracking.cpp [hunk 3]
```diff
@@ -37,6 +38,11 @@ exprt find_object_base_expression(const address_of_exprt &address_of)
   return current.get();
 }
 
+exprt find_object_base_expression(const pointer_object_exprt &pointer_object)
+{
+  return pointer_object.pointer();
```
Comment (thomasspriggs) [inline src/solvers/smt2_incremental/object_tracking.cpp]: :no_entry_sign: The result of this function you have added is a pointer expression, not an "object base". Therefore it should not be added to the object map. We should probably discuss this outside of the review comments to go over the logic for this. The definition of the null object is kind of the exception to the rule which could be misleading.

:information_source: The pointer expression could be expressions such as `&x` or `(&x) + 10`. Neither of which should be added to the map. The map will already have the `x` expression required added to it. New entries should not be added for these other expressions as `x` is the base object of both of them. These pointer expressions should not be given new unique object identifiers as they do not refer to distinct objects.

:information_source: The pointer expression could contain multiple object base expressions in cases that look like - `pointer_object(nondet_bool ? (&x) : (&y))`

File: src/solvers/smt2_incremental/convert_expr_to_smt.cpp [hunk 4]
```diff
@@ -1142,6 +1142,28 @@ static smt_termt convert_expr_to_smt(
     mult_overflow.pretty());
 }
 
+static smt_termt convert_expr_to_smt(
+  const pointer_object_exprt &pointer_object,
+  const sub_expression_mapt &converted,
+  const smt_object_mapt &object_map)
+{
+  const auto base = pointer_object.pointer();
```
Comment (thomasspriggs) [inline src/solvers/smt2_incremental/convert_expr_to_smt.cpp]: Why have you called this pointer expression `base`?

File: src/solvers/smt2_incremental/convert_expr_to_smt.cpp [hunk 5]
```diff
@@ -1142,6 +1142,28 @@ static smt_termt convert_expr_to_smt(
...
+  const auto object = object_map.find(base);
+  INVARIANT(
+    object != object_map.end(),
+    "Objects should be tracked before converting their object ID to SMT terms");
+  const std::size_t object_id = object->second.unique_id;
+  INVARIANT(
+    object_id < power2(config.bv_encoding.object_bits),
+    "There should be sufficient bits to encode unique object identifier.");
+  const smt_termt object_bit_vector =
+    smt_bit_vector_constant_termt{object_id, config.bv_encoding.object_bits};
```
Comment (thomasspriggs) [inline src/solvers/smt2_incremental/convert_expr_to_smt.cpp]: :no_entry_sign: We cannot guarantee that the pointer a `pointer_object_exprt` is applied to will always point to the same constant object. Therefore the logic you have used here must be incorrect. I am going do some manual testing to put together and example where what you have implemented does not work as it should.

**Review by thomasspriggs — CHANGES_REQUESTED** [review]

**Comment by thomasspriggs** [comment]
The SAT back end can produce a counter example for the following input. Whereas the SMT back end in your branch cannot -
```
#include <assert.h>
#include <stdbool.h>

int main()
{
	int x;
	int y;
	int z;
	bool nondet1;
	bool nondet2;
	int *a = nondet1 ? &x : &y;
	int *b = nondet2 ? &y : &z;
	__CPROVER_assert(!__CPROVER_same_object(a, b), "Can be violated.");
}
```

File: src/solvers/smt2_incremental/smt_bit_vector_theory.cpp (hunk 2 above)
Comment (thomasspriggs) [reply src/solvers/smt2_incremental/smt_bit_vector_theory.cpp]: Looks to have been sorted.

File: regression/cbmc-incr-smt2/pointers/pointers_simple.c [hunk 6]
```diff
@@ -0,0 +1,14 @@
+int main()
+{
+  int a = 10;
+  int *b = &a;
+  int c;
+
+  *b = 12;
+
+  __CPROVER_assert(a != *b, "a should be different than b");
```
Comment (thomasspriggs) [inline regression/cbmc-incr-smt2/pointers/pointers_simple.c:9]: :pick: The title of this commit suggests it is about pointer objects but it looks to be just a test of dereferencing pointers to integers?

File: src/solvers/smt2_incremental/convert_expr_to_smt.cpp [hunk 7]
```diff
@@ -1142,6 +1142,28 @@ static smt_termt convert_expr_to_smt(
...
+static smt_termt convert_expr_to_smt(
+  const pointer_object_exprt &pointer_object,
+  const sub_expression_mapt &converted)
+{
+  const auto type =
+    type_try_dynamic_cast<bitvector_typet>(pointer_object.type());
+  INVARIANT(type, "Pointer object should have a bitvector-based type.");
+  const auto pointer_expr = pointer_object.pointer();
+  const auto converted_expr = converted.at(pointer_expr);
+  const std::size_t pointer_width = type->get_width();
```
Comment (thomasspriggs) [inline src/solvers/smt2_incremental/convert_expr_to_smt.cpp]: :exclamation: `pointer_width` is misnamed, because it is not necessarily in the width of a pointer. It is the width of the result of the pointer_object expression which is just an integer, not a pointer.

File: src/solvers/smt2_incremental/convert_expr_to_smt.cpp [hunk 8]
```diff
@@ -1142,6 +1142,28 @@ static smt_termt convert_expr_to_smt(
...
 }
 
+static smt_termt convert_expr_to_smt(
+  const pointer_object_exprt &pointer_object,
+  const sub_expression_mapt &converted)
+{
+  const auto type =
+    type_try_dynamic_cast<bitvector_typet>(pointer_object.type());
+  INVARIANT(type, "Pointer object should have a bitvector-based type.");
+  const auto pointer_expr = pointer_object.pointer();
```
Comment (thomasspriggs) [inline src/solvers/smt2_incremental/convert_expr_to_smt.cpp]: :pick: I suggest just calling `pointer_object.pointer()` inside the call to `converted.at`, rather than giving it a new name/variable. This is because the new variable is only used once at I don't find the extra name to be that much more informative.

File: src/solvers/smt2_incremental/convert_expr_to_smt.cpp [hunk 9]
```diff
@@ -1142,6 +1142,28 @@ static smt_termt convert_expr_to_smt(
...
+  const auto converted_expr = converted.at(pointer_expr);
+  const std::size_t pointer_width = type->get_width();
+  INVARIANT(pointer_width > 0, "Pointer should have a positive width.");
```
Comment (thomasspriggs) [inline src/solvers/smt2_incremental/convert_expr_to_smt.cpp]: We should check `pointer_width >= object_bits_width` rather than `pointer_width > 0` because that would be the more stringent check and it is needed in order to confirm that `pointer_width - object_bits_width` does not overflow / is greater than 0.

File: src/solvers/smt2_incremental/convert_expr_to_smt.cpp [hunk 10]
```diff
@@ -1142,6 +1142,28 @@ static smt_termt convert_expr_to_smt(
...
+  const std::size_t object_bits_width = config.bv_encoding.object_bits;
+  const std::size_t ext = pointer_width - object_bits_width;
+  const auto extract_op = smt_bit_vector_theoryt::extract(
+    pointer_width - 1, pointer_width - object_bits_width)(converted_expr);
+  if(ext > 0)
```
Comment (thomasspriggs) [inline src/solvers/smt2_incremental/convert_expr_to_smt.cpp:1161]: :grey_question: Is it reasonable for this to be an invariant rather than a different code path? That might be a simpler implementation.

File: src/solvers/smt2_incremental/construct_value_expr_from_smt.cpp [hunk 11]
```diff
@@ -37,13 +37,32 @@ class value_expr_from_smt_factoryt : public smt_term_const_downcast_visitort
...
+      INVARIANT(
+        pointer_type->get_width() == sort_width,
+        "Width of smt bit vector term must match the width of pointer  type.");
+      if(bit_vector_constant.value() == 0)
+      {
+        result = null_pointer_exprt{*pointer_type};
+      }
+      else
+      {
+        result = constant_exprt{
```
Comment (thomasspriggs) [inline src/solvers/smt2_incremental/construct_value_expr_from_smt.cpp]: :no_entry_sign: As we had previously discussed, this commit should be a separate PR with its own tests. It should not really just be bundled in with other work.

File: regression/cbmc-incr-smt2/pointers/pointer_object.c [hunk 12]
```diff
@@ -0,0 +1,20 @@
...
+  int nondet_bool;
+  int flag = 1;
+
+  int *b = &a;
+  int *c = nondet_bool ? &a : 0;
+  int *d = flag ? &a : 0;
+  int *e;
+
+  // __CPROVER_same_object is True when
+  // `__CPROVER_pointer_object(a) == __CPROVER_pointer_object(b)`
```
Comment (thomasspriggs) [inline regression/cbmc-incr-smt2/pointers/pointer_object.c:13]: :no_entry_sign: This PR is missing regression tests where `__CPROVER_pointer_object(a)` is used directly. The object identifiers are allocated deterministically. So you should be able to write tests based on the following properties -

- `__CPROVER_pointer_object(NULL)` is zero.
- `__CPROVER_pointer_object(&foo)` is one in the case where a program contains a single address of operation.

File: unit/solvers/smt2_incremental/convert_expr_to_smt.cpp [hunk 13]
```diff
@@ -1186,3 +1187,57 @@ TEST_CASE(
...
+  "expr to smt conversion for pointer object expression",
+  "[core][smt2_incremental]")
+{
+  const symbol_exprt foo{"foo", pointer_typet(unsigned_int_type(), 64)};
+  const symbol_exprt foobar{"foobar", pointer_typet(unsigned_int_type(), 64)};
+
+  SECTION("Pointer object expression")
+  {
+    const pointer_object_exprt pointer_obj =
+      *expr_try_dynamic_cast<pointer_object_exprt>(pointer_object(foo));
```
Comment (thomasspriggs) [inline unit/solvers/smt2_incremental/convert_expr_to_smt.cpp]: :pick: If you are going to immediately dereference the result of `expr_try_dynamic_cast`, you should use `expr_checked_cast` instead.
:no_entry_sign: `pointer_object` depends on `size_type`, which depends on the global configuration, which isn't being set in this test. That means that this test could fail if the preceding unit tests executed in the same run have set the configuration differently than this test expects.
:pick: This cast is superfluous and should be removed. If you want to make the type explicit, you could just construct `pointer_object_exprt` directly instead of using `pointer_object`.


This comment also applies to other lines in this commit using the free `pointer_object` function.

File: regression/cbmc-incr-smt2/pointers/pointer_object2.desc [hunk 14]
```diff
@@ -0,0 +1,11 @@
...
+pointer_object2.c
+--trace --verbosity 10
+\[main\.assertion\.1\] line 13 Can be violated.: FAILURE
+nondet1=FALSE
+nondet2=TRUE
+^VERIFICATION FAILED$
+^EXIT=10$
+^SIGNAL=0$
+--
+--
```
Comment (thomasspriggs) [inline regression/cbmc-incr-smt2/pointers/pointer_object2.desc:11]: :pick: Missing a test description.

File: src/solvers/smt2_incremental/convert_expr_to_smt.cpp [hunk 15]
```diff
@@ -1164,6 +1164,29 @@ static smt_termt convert_expr_to_smt(
...
+static smt_termt convert_expr_to_smt(
+  const pointer_offset_exprt &pointer_offset,
+  const sub_expression_mapt &converted)
+{
+  const auto type =
+    type_try_dynamic_cast<bitvector_typet>(pointer_offset.type());
+  INVARIANT(type, "Pointer offset should have a bitvector-based type.");
+  const auto pointer_expr = pointer_offset.pointer();
+  const auto converted_expr = converted.at(pointer_expr);
+  const std::size_t pointer_width = type->get_width();
```
Comment (thomasspriggs) [inline src/solvers/smt2_incremental/convert_expr_to_smt.cpp]: :pick: This is the width of the result of the operation, not the width of a pointer. Hence it is misnamed.

File: src/solvers/smt2_incremental/convert_expr_to_smt.cpp [hunk 16]
```diff
@@ -1164,6 +1164,29 @@ static smt_termt convert_expr_to_smt(
...
+  std::size_t offset_bits = pointer_width - config.bv_encoding.object_bits;
+  if(offset_bits > pointer_width)
+    offset_bits = pointer_width;
+  const auto extract_op =
+    smt_bit_vector_theoryt::extract(offset_bits - 1, 0)(converted_expr);
+  if(pointer_width > offset_bits)
+  {
+    return smt_bit_vector_theoryt::zero_extend(pointer_width - offset_bits)(
```
Comment (thomasspriggs) [inline src/solvers/smt2_incremental/convert_expr_to_smt.cpp]: :thinking: The result type of `pointer_offset_exprt` is usually signed. Therefore we should probably be using `sign_extend` instead of `zero_extend`. Note that the old SMT2 back end may may wrong in this case. Note also that the signed-ness is currently set in `pointer_predicates.cpp` but we have no guarantees that these expressions will always have signed/unsigned results (unless we add appropriate invariants).
:information_source: Negative offsets may occur in the the case of pointer offset overflows.

File: src/solvers/smt2_incremental/convert_expr_to_smt.cpp [hunk 17]
```diff
@@ -1164,6 +1164,29 @@ static smt_termt convert_expr_to_smt(
...
+  const sub_expression_mapt &converted)
+{
+  const auto type =
+    type_try_dynamic_cast<bitvector_typet>(pointer_offset.type());
+  INVARIANT(type, "Pointer offset should have a bitvector-based type.");
+  const auto pointer_expr = pointer_offset.pointer();
+  const auto converted_expr = converted.at(pointer_expr);
+  const std::size_t pointer_width = type->get_width();
+  std::size_t offset_bits = pointer_width - config.bv_encoding.object_bits;
+  if(offset_bits > pointer_width)
```
Comment (thomasspriggs) [inline src/solvers/smt2_incremental/convert_expr_to_smt.cpp]: :pick: Are we actually truncating the result in this case? Might be worth adding a comment to that effect as this is slightly surprising and it might be worth just adding an invariant to avoid surprise truncation instead.

File: src/solvers/smt2_incremental/convert_expr_to_smt.cpp [hunk 18]
```diff
@@ -1164,6 +1164,29 @@ static smt_termt convert_expr_to_smt(
   return extract_op;
 }
 
+static smt_termt convert_expr_to_smt(
```
Comment (thomasspriggs) [inline src/solvers/smt2_incremental/convert_expr_to_smt.cpp:1168]: :question: Where are the regression tests for this conversion?

File: unit/solvers/smt2_incremental/convert_expr_to_smt.cpp [hunk 19]
```diff
@@ -1241,3 +1241,35 @@ TEST_CASE(
...
 }
+
+TEST_CASE("pointer_offset_exprt to SMT conversion", "[core][smt2_incremental]")
+{
+  const symbol_exprt foo{"foo", pointer_typet(unsigned_int_type(), 64)};
+
+  SECTION("simple pointer_offset_exprt conversion")
+  {
+    const pointer_offset_exprt pointer_offset_exp =
+      *expr_try_dynamic_cast<pointer_offset_exprt>(pointer_offset(foo));
```
Comment (thomasspriggs) [inline unit/solvers/smt2_incremental/convert_expr_to_smt.cpp]: See previous comments on `*expr_try_dynamic_cast` and `pointer_offset` in the other unit test.

**Review by thomasspriggs — CHANGES_REQUESTED** [review]

File: src/solvers/smt2_incremental/convert_expr_to_smt.cpp (hunk 7 above)
Comment (NlightNFotis) [reply src/solvers/smt2_incremental/convert_expr_to_smt.cpp]: Renamed to just `width` and the `object_bit_width` to just `object_bits` since it's acting as just an alias of the other definition.

Comment (NlightNFotis) [reply src/solvers/smt2_incremental/convert_expr_to_smt.cpp]: Worth noting here that the old SMT backend does call the same computation `pointer_width`

File: src/solvers/smt2_incremental/convert_expr_to_smt.cpp (hunk 10 above)
Comment (NlightNFotis) [reply src/solvers/smt2_incremental/convert_expr_to_smt.cpp:1161]: I don't think so.

I modelled this after the old SMT backend, and there it's encoded in much the same way, which seems to imply that `ext` can be at least equal to `0`, in which case we don't do the `zero_extend`.

File: src/solvers/smt2_incremental/construct_value_expr_from_smt.cpp (hunk 11 above)
Comment (NlightNFotis) [reply src/solvers/smt2_incremental/construct_value_expr_from_smt.cpp]: Left here for reference: 

This is submitted as its own PR in https://github.com/diffblue/cbmc/pull/6831

File: unit/solvers/smt2_incremental/convert_expr_to_smt.cpp (hunk 1 above)
Comment (thomasspriggs) [reply unit/solvers/smt2_incremental/convert_expr_to_smt.cpp:1195]: Looks like this still needs fixing.

File: regression/cbmc-incr-smt2/pointers/pointer_offset.c [hunk 20]
```diff
@@ -0,0 +1,16 @@
...
+
+  __CPROVER_assert(
+    __CPROVER_POINTER_OFFSET(p) != __CPROVER_POINTER_OFFSET(q),
+    "expected failure because offsets should be the same");
+
+  // TODO: Remove comments once pointer arithmetic works:
+
+  // *q = p + 2;
+
+  // __CPROVER_assert(__CPROVER_POINTER_OFFSET(p) != __CPROVER_POINTER_OFFSET(q), "expected failure");
```
Comment (thomasspriggs) [inline regression/cbmc-incr-smt2/pointers/pointer_offset.c:15]: :pick: This assertion looks like it should pass once we have support for integer arithmetic. The assignment above will give `q` an offset of `2 * sizeof(int)` and p will still have an offset of zero.

**Review by thomasspriggs — APPROVED** [review]
